File: PRACTICA6/PYTHON/AlgebraVectorial2.py
import math

class Vector:
    def __init__(self, x=0, y=0, z=0):
        self.x = x
        self.y = y
        self.z = z

    def add(self, other):
        return Vector(self.x + other.x, self.y + other.y, self.z + other.z)

    def multiply(self, other):
        if isinstance(other, (int, float)):
            return Vector(self.x * other, self.y * other, self.z * other)
        elif isinstance(other, Vector):
            dot_product = self.x * other.x + self.y * other.y + self.z * other.z
            return "Un vector ortogonal" if dot_product == 0 else dot_product
        return None

    def cross(self, other):
        return Vector(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x
        )

    def length(self):
        return math.sqrt(self.x ** 2 + self.y ** 2 + self.z ** 2)

    def normalize(self):
        len = self.length()
        return Vector() if len == 0 else Vector(self.x / len, self.y / len, self.z / len)

    def __str__(self):
        return f"(x={self.x:.2f}, y={self.y:.2f}, z={self.z:.2f})"

# Test
if __name__ == "__main__":
    v1 = Vector(5, 4, 2)
    v2 = Vector(-2, 3, -1)

    print(v1)
    print(v2)

    v3 = v1.add(v2)
    print("El vector suma es", v3)

    v4 = v1.multiply(3)
    print("El vector multiplicado por escalar es", v4)

File: PRACTICA6/JAVA/AlgebraVectorial2.java
public class Vector {
    private double x, y, z;

    // Constructor
    public Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    // Constructor por defecto
    public Vector() {
        this(0, 0, 0);
    }

    // Suma de vectores
    public Vector add(Vector other) {
        return new Vector(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    // Multiplicación por escalar o producto punto
    public Object multiply(Object other) {
        if (other instanceof Number) {
            double scalar = ((Number) other).doubleValue();
            return new Vector(this.x * scalar, this.y * scalar, this.z * scalar);
        } else if (other instanceof Vector) {
            Vector v = (Vector) other;
            double dotProduct = this.x * v.x + this.y * v.y + this.z * v.z;
            return dotProduct == 0 ? "Un vector ortogonal" : dotProduct;
        }
        return null;
    }

    // Producto cruzado
    public Vector cross(Vector other) {
        return new Vector(
            this.y * other.z - this.z * other.y,
            this.z * other.x - this.x * other.z,
            this.x * other.y - this.y * other.x
        );
    }

    // Longitud del vector
    public double length() {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }

    // Normalizar
    public Vector normalize() {
        double len = length();
        return len == 0 ? new Vector() : new Vector(this.x / len, this.y / len, this.z / len);
    }

    @Override
    public String toString() {
        return String.format("(x=%.2f, y=%.2f, z=%.2f)", x, y, z);
    }

    // Test
    public static void main(String[] args) {
        Vector v1 = new Vector(5, 4, 2);
        Vector v2 = new Vector(-2, 3, -1);

        System.out.println(v1);
        System.out.println(v2);

        Vector v3 = v1.add(v2);
        System.out.println("El vector suma es " + v3);

        Vector v4 = (Vector) v1.multiply(3);
        System.out.println("El vector multiplicado por escalar es " + v4);
    }
}
